### ISSUE TITLE
Config change  request for tickless kernel

### ISSUE THREAD

**Issue by linuxelf001** [microsoft/WSL2-Linux-Kernel#2]
CONFIG_NO_HZ_FULL is not set
CONFIG_NO_HZ is not set

Not that ticks are much load but to minimize/zero CPU load of having it idle in the background, can CONFIG_NO_HZ be enabled?

**Comment by kdrag0n** [comment]
Tickless idle (`CONFIG_NO_HZ_IDLE`) [is enabled](https://github.com/microsoft/WSL2-Linux-Kernel/blob/master/Microsoft/config-wsl#L78), which significantly reduces the background CPU load. It's already almost consistently zero when not being used.

Full tickless (`CONFIG_NO_HZ_FULL`) can have a substantial impact on overall performance outside of HPC workloads, so it probably wouldn't be of much use to the majority of WSL 2 users. That being said, it has received gradual improvements throughout the kernel versions so it might be worthy of another chance on 4.19. I wouldn't count on it though, because no major Linux distributions have enabled it in their official kernel.

Despite the old advice, preliminary tests of a full tickless kernel running on nested WSL 2 (with a custom 4.19.56 kernel) have been fairly promising with regards to performance:

| Test | Time on tickless idle | Time on full tickless |
| --- | --- | --- |
| `clock_gettime(CLOCK_MONOTONIC)` syscall | 408 ns | 403 ns |
| `clock_gettime(CLOCK_MONOTONIC)` vDSO call | 18 ns | 19 ns |
| `mmap` I/O calls on procfs | 2283 ns | 2297 ns |
| POSIX I/O calls on procfs | 2949 ns | 2955 ns |
| `hackbench -p 20 -T 1000 -l 1000` | 2.038 s | 1.946 s |

`CONFIG_NO_HZ` is a legacy option that is only kept around to preserve compatibility with old configurations and is slated for removal in the near future.

**Comment by kdrag0n** [comment]
It doesn't look like `CONFIG_NO_HZ_FULL` will help in the intended way on WSL 2. I measured the effective timer interrupt (Hyper-V `stimer0` synthetic timer) rate and found that it was approximately 8.5 Hz on tickless idle and 12 Hz on full tickless — a **41% regression**.

Below is the script I used to measure the interrupt rate:
```bash
old=0

while true
do
	new=$(cat /proc/interrupts | grep HVS | awk '{print $2}')
	diff=$((new - old))
	echo $diff

	old=$new
	sleep 1
done
```

For what it's worth, I also measured the time taken to compile a slightly-modified WSL 2 kernel in it and found basically no difference. It took 3m36.28s on full tickless and 3m36.83s on tickless idle.

Boot times were pretty much the same story, if not slightly in favor of tickless idle: 971.79 ms on tickless idle and 996.04 ms on full tickless, averaged across 10 boots. Here's the PowerShell script I used for this if anyone wants to replicate my results: https://gist.github.com/kdrag0n/2de3a51bb2aac99cfb9003534bed002e

Overall, I think tickless idle is still the best option for WSL as performance is nearly identical and the idle timer interrupt rate is lower.
